### ISSUE TITLE
package a docker image

### ISSUE THREAD

**Issue by g-k** [mozilla/frost#302]
It's weird that we:

* have both a repo and the pypi package: https://pypi.org/project/frost/ and 
* create a virtualenv for the install (e.g. in docker images)

We could run `frost` as an alias to `pytest` on a directories of tests or individual test files with an optional config file (and other pytest args passing through to pytest):

`frost --config config.yaml tests/`

It'd also let us bundle up tests and their related test resources as pytest plugins in the future: https://docs.pytest.org/en/stable/writing_plugins.html#writing-plugins

**Comment by g-k** [comment]
f? @ajvb

**Comment by ajvb** [comment]
@g-k I completely agree. I also think this would help with usability and user experience for those new to frost. Saying "Run the `frost` command" is a lot easier than saying "Run the `pytest` command with this selection of custom pytest cli args but ignore all these other ones" imo

**Comment by g-k** [comment]
Renaming issue to package as a docker image. I was running into issues with outdated pytest and generally wrapping it without breaking existing APIs.

I think we can:

- [ ] package frost as a docker image
- [ ] build and publish it from CI (need creds from a mozilla docker admin)
- [ ] migrate existing users to the docker image
- [ ] then eventually swap the CLI if we want

The benefits would be:

* pin our Python and pytest versions
* default to not running all tests
* mount in user tests
* add extra non-python tools and not worry about platform specific changes

An additional consideration is getting this working nicely with VSCode since Hal and other users use it.

**Comment by ajvb** [comment]
@g-k In regards to trouble with wrapping pytest, do you know if that would be resolved with upgrading pytest? I still really like the user experience story of having a `frost` command, even if we just keep it on the backlog for a while.

**Comment by g-k** [comment]
I did run into the `ignore_tests` config requiring a newer version, but didn't want to try to update pytest with the CLI change.

I was able to run pytest programmatically and pass through args using the click packages, so I'll probably revisit that at some point. Ideally we make frost less magical, remove reporting and other plugins, weird conftest stuff, and just wrap the individual test running functionality sort of like:

```console
$ cat jsonlines-data.jsonl | front this-one-weird--included-test > output.jsonl # where output is a report or a list of actions to run in other tools (create bugs, file issues, etc.)
$ cat jsonlines-data.jsonl | front my-test.py > output.jsonl
$ # or run multiple tests
$ frost --config my-config-specifying-tests-to-run
```

**Comment by g-k** [comment]
The docker image is busted.

I have some fixes in https://github.com/mozilla/frost/pull/313 (we do need to `COPY` in each dir but it's gross and less maintainable. Moving everything into a `frost/` dir as in https://github.com/mozilla/frost/pull/303 would make it easier to package).

Another option @smarnach mentioned (intermediate or otherwise) is using tox. It'll set up and run things in the virtualenv, which would be quite nice to:

* reduce complexity in the Makefile
* simplify initial install

A somewhat newer option is nox https://nox.thea.codes/en/stable/ which would give us programmatic install.

@ajvb I think the larger question is: where do we want frost to go and how much breaking change is acceptable on master? Right now the impact of getting things running with a different CLI looks like:

* upgrade pytest a couple major versions
* move everything to a top level `frost/` dir
* add a `setup.py`
* change the entrypoint to a custom CLI or docker image (the docker image is somewhat orthogonal though)

I'm wondering whether it would be better to work in a 2.0 branch.

**Comment by ajvb** [comment]
>Moving everything into a frost/ dir as in #303 would make it easier to package

That sounds good to me.

> Another option @smarnach mentioned (intermediate or otherwise) is using tox. It'll set up and run things in the virtualenv, which would be quite nice to:
> 
>     * reduce complexity in the Makefile
> 
>     * simplify initial install
> 
> 
> A somewhat newer option is nox https://nox.thea.codes/en/stable/ which would give us programmatic install.

I'm not against using nox, but would want it as an intermediate to either using a docker container or a `frost` cli command.
 
> @ajvb I think the larger question is: where do we want frost to go and how much breaking change is acceptable on master? Right now the impact of getting things running with a different CLI looks like:
> 
>     * upgrade pytest a couple major versions
> 
>     * move everything to a top level `frost/` dir
> 
>     * add a `setup.py`
> 
>     * change the entrypoint to a custom CLI or docker image (the docker image is somewhat orthogonal though)
> 
> 
> I'm wondering whether it would be better to work in a 2.0 branch.

I think that breaking changes are acceptable, but yeah, I agree that it might be better to start working on a 2.0 branch. I'd also like to suggest that we do some initial planning/project management on what all we want to do here around making frost easier to use. How do we want to do releases generally? Do we want to start pushing up a package to PyPI?

To your question about "where do we want frost to go", I think most of my thought around this stuff is generally making it easier to use. I think right now frost is confusing for new users whether or not you are familiar with pytest. It might actually be harder for those who are used to pytest lol as we've built our own framework around it.

tl;dr I want users to be able to have a one-liner to install `frost` and to be able to do stuff like `frost --help` and get just the important and needed CLI args.

@g-k How does that all sound?
